### PR TITLE
Added support of collection initializer syntax.

### PR DESCRIPTION
**Pull request by holdenmai**
Implementation of collection initializer syntax support for constructor #194   Managed to somehow close the original PR on accident during rebase operation [pull 246](https://github.com/dynamicexpresso/DynamicExpresso/pull/246), but all comments have now been resolved.

You can now write

```
var target = new Interpreter();
target.Reference(typeof(System.Collections.Generic.List<>));
var intList = target.Eval<System.Collections.Generic.List<int>>("new List<int>(){1, 2, 3, 4, 5}");
Assert.AreEqual(5, intList.Count);
for (int i = 0; i < intList.Count; ++i)
{
    Assert.AreEqual(i + 1, intList[i]);
}
or 
 target.Eval<System.Collections.Generic.List<string>>("new List<string>(){string.Empty}");
```

as well as implementations supporting multiple parameters in the Add method such as 

`new MyClassAdder(){{ 1, 2, 3, 4, 5},{\"6\" },7	}`
and
`new Dictionary<int, string>(){{1, \"1\"}, {2, \"2\"}, {3, \"3\"}, {4, \"4\"}, {5, \"5\"}}`

Reference to method parameters is also supported like below

`target.Eval<System.Collections.Generic.List<string>>("new List<string>(){StrProp = string.Empty}", new Parameter("StrProp", "0"))
target.Eval<System.Collections.Generic.List<string>>("new List<string>(){StrValue()}", new Parameter("StrValue", new Func<string>(() => "Func")))`

### CODE REVIEW

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 1]
```diff
@@ -1207,41 +1207,126 @@ private Expression[] ParseArrayInitializerList()
...
+			return initializedInstance;
 		}
 
-		private MemberBinding[] ParseMemberInitializerList(Type newType)
+		private Expression ParseMemberAndInitializerList(NewExpression newExpr, Type newType)
 		{
+			var originalPos = _token.pos;
 			var bindingList = new List<MemberBinding>();
+			var actions = new List<Expression>();
+			ParameterExpression instance = Expression.Variable(newType);
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:1225]: Please  always use var :)

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs [hunk 2]
```diff
@@ -179,6 +179,84 @@ public void Array_multi_dimension_constructor()
...
+			}
+			Assert.AreEqual("6", l.StrProp);
+			Assert.AreEqual(7, l.IntField);
+		}
+
+
+		[Test]
+		public void Ctor_NewMyClassWithCross()
+		{
+			string StrProp = "";
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ConstructorTest.cs:214]: Please always use var, and use `string.Empty`  insteaf of `""`

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs [hunk 3]
```diff
@@ -179,6 +179,84 @@ public void Array_multi_dimension_constructor()
...
+			var mc = new MyClassAdder()
+			{
+				{1,2,3 },
+				{StrProp = "6" }
+			};
+			if (StrProp == "6")
+			{
+
+			}
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ConstructorTest.cs:223]: Can be removed

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs [hunk 4]
```diff
@@ -179,6 +179,84 @@ public void Array_multi_dimension_constructor()
...
+			var target = new Interpreter();
+			target.Reference(typeof(MyClassAdder));
+			Assert.DoesNotThrow(() => target.Eval<MyClassAdder>("new MyClassAdder(){{ 1, 2, 3, 4, 5},{StrProp = \"6\" },7}", new Parameter("StrProp", "0")));
+			Assert.DoesNotThrow(() => target.Eval<MyClassAdder>("new MyClassAdder(){{ 1, 2, 3, 4, 5},StrProp = \"6\",7}", new Parameter("StrProp", "0")));
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ConstructorTest.cs:227]: If it's not expected to throw, can you instead test the expected results? In particular, that the expected value of StrProp is "6"?

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs [hunk 5]
```diff
@@ -179,6 +179,84 @@ public void Array_multi_dimension_constructor()
...
+			var target = new Interpreter();
+			target.Reference(typeof(System.Collections.Generic.List<>));
+			var list = target.Eval<System.Collections.Generic.List<string>>("new List<string>(){string.Empty}");
+			Assert.AreEqual(1, list.Count);
+			for (int i = 0; i < list.Count; ++i)
+			{
+				Assert.AreSame(string.Empty, list[i]);
+			}
+			Assert.DoesNotThrow(() => target.Eval<System.Collections.Generic.List<string>>("new List<string>(){StrProp = string.Empty}", new Parameter("StrProp", "0")));
+			Assert.DoesNotThrow(() => target.Eval<System.Collections.Generic.List<string>>("new List<string>(){StrValue()}", new Parameter("StrValue", new Func<string>(() => "Func"))));
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ConstructorTest.cs:256]: Please check the expected values.

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs [hunk 6]
```diff
@@ -179,6 +179,84 @@ public void Array_multi_dimension_constructor()
...
+			Assert.DoesNotThrow(() => target.Eval<MyClassAdder>("new MyClassAdder(){{ 1, 2, 3, 4, 5},StrProp = \"6\",7}", new Parameter("StrProp", "0")));
+			Assert.DoesNotThrow(() => target.Eval<MyClassAdder>("new MyClassAdder(){{ 1, 2, 3, 4, 5},string.Empty, 7}"));
+			Assert.Throws<ParseException>(() => target.Eval<MyClassAdder>("new MyClassAdder(){{ 1, 2, 3, 4, 5},{StrProp = \"6\" },7	}"));
+		}
+		[Test]
+		public void Ctor_NewListWithItems()
+		{
+			var target = new Interpreter();
+			target.Reference(typeof(System.Collections.Generic.List<>));
+			var intList = target.Eval<System.Collections.Generic.List<int>>("new List<int>(){1, 2, 3, 4, 5}");
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ConstructorTest.cs:236]: Can you add a test with a property access? For example, replace 5 with `int.MaxValue`, or a parameter? To ensure you can use any expression (I saw the next test uses a method call, but one more test won't hurt).

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 7]
```diff
@@ -1207,41 +1207,126 @@ private Expression[] ParseArrayInitializerList()
...
+						this._arguments.TryGetIdentifier(_token.text, out _)
+							||
+						FindPropertyOrField(newType, _token.text, false) == null)
+				{
+					ParseCollectionInitalizer(newExpr, newType, originalPos, bindingList, actions, instance, allowCollectionInit);
+					return;
+				}
+				else
+				{
+					throw CreateParseException(originalPos, ErrorMessages.InvalidInitializerMemberDeclarator);
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:1266]: Can you please add a unit test for this exception?

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 8]
```diff
@@ -1207,41 +1207,126 @@ private Expression[] ParseArrayInitializerList()
...
-				if (_token.id != TokenId.Comma) break;
+			var value = ParseExpressionSegment();
+			bindingList.Add(Expression.Bind(member, value));
+		}
+
+		private void ParseCollectionInitalizer(NewExpression newExpr, Type newType, int originalPos, List<MemberBinding> bindingList, List<Expression> actions, ParameterExpression instance, bool allowCollectionInit)
+		{
+			if (!allowCollectionInit)
+			{
+				throw CreateParseException(_token.pos, ErrorMessages.CollectionInitializationNotSupported, newType, typeof(IEnumerable));
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:1303]: Please add a unit test for this.

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 9]
```diff
@@ -1207,41 +1207,126 @@ private Expression[] ParseArrayInitializerList()
...
+			}
+			
+			if (bindingList.Count > 0)
+			{
+				throw CreateParseException(originalPos, ErrorMessages.InvalidInitializerMemberDeclarator);
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:1308]: Please add a unit test for this.

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 10]
```diff
@@ -1207,41 +1207,126 @@ private Expression[] ParseArrayInitializerList()
...
+			{
+				var pos = _token.pos;
+				if (allowCollectionInit)
+				{
+					NextToken();
+					if (_token.id != TokenId.Equal || _arguments.TryGetIdentifier(propertyOrFieldName, out _) || _arguments.TryGetParameters(propertyOrFieldName, out _))
+					{
+						SetTextPos(pos);
+						NextToken();
+						actions.Add(ParseNormalMethodInvocation(newType, instance, _token.pos, "Add", new[] { ParseExpressionSegment() }));
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:1283]: Should call `ParseCollectionInitalizer`

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 11]
```diff
@@ -1207,41 +1207,126 @@ private Expression[] ParseArrayInitializerList()
...
+			if (_token.id == TokenId.OpenCurlyBracket)
+			{
+				var pos = _token.pos;
 				NextToken();
+				if (_token.id == TokenId.Identifier)
+				{
+					if (!_arguments.TryGetIdentifier(_token.text, out var _) && !_arguments.TryGetParameters(_token.text, out var _))
+					{
+						throw CreateParseException(_token.pos, ErrorMessages.InvalidInitializerMemberDeclarator);
+					}
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:1319]: I think this will prevent class constants from being used, such as `string.Empty`.

**Review by metoule — COMMENTED** [review]
That looks really good! Apart from a few nitpicks, I'd recommend adding more tests, mostly to test the various possible exceptions. Also, I believe the current implementation can't parse `new List<string> { { string.Empty } }`.

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs (hunk 2 above)
Comment (holdenmai) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs]: Incorporated in latest update

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs (hunk 3 above)
Comment (holdenmai) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs]: Incorporated in latest update

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs (hunk 5 above)
Comment (holdenmai) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs:294]: Added more positive and negative testing around generic list creation.

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 10 above)
Comment (holdenmai) [reply src/DynamicExpresso.Core/Parsing/Parser.cs]: Cleaned up the logic around transitioning over to collection init from member init process.

**Comment by holdenmai** [comment]
I have added scenarios within the differing unit tests to cover each of the cases mentioned.  Is it preferred that they become standalone cases to better show what negative case they are trying to cover?

**Comment by davideicardi** [comment]
Thank you @holdenmai 

> Is it preferred that they become standalone cases to better show what negative case they are trying to cover?
Yes, maybe having separate tests would be better... or try to simplify/explain better the existing ones.

**Comment by holdenmai** [comment]
I have reorganized the unit tests to better show the different scenarios being covered.

**Comment by davideicardi** [comment]
LGTM!

@metoule ? 

There is just 3 conversations still open that I'm not sure if they are now resolved ...

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 8 above)
Comment (holdenmai) [reply src/DynamicExpresso.Core/Parsing/Parser.cs:1296]: Covered by Ctor_CannotUseCollectionInitDoesNotImplementIEnumerable test

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 9 above)
Comment (holdenmai) [reply src/DynamicExpresso.Core/Parsing/Parser.cs:1300]: covered in multiple tests under Ctor_InvalidInitializerMemberDeclarator test

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 11 above)
Comment (holdenmai) [reply src/DynamicExpresso.Core/Parsing/Parser.cs]: The scenario of class constants has been checked in a couple different tests.
See Ctor_NewMyClass_ExpectedValues and Ctor_NewListWithItems for validation of different ways of referencing those values.

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs (hunk 6 above)
Comment (davideicardi) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs]: @holdenmai I think this is the last missing test. Or it is already done?

Comment (holdenmai) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs]: the Ctor_NewListWithItems has a lot of different tests calling out to different "constants", both direct and referencing out to int.MaxValue.

and Ctor_NewListWithString also contains references to using parameters.

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 12]
```diff
@@ -1207,41 +1207,134 @@ private Expression[] ParseArrayInitializerList()
...
+				{
+					if (actions.Count > 0)
+					{
+						throw CreateParseException(pos, ErrorMessages.InvalidInitializerMemberDeclarator);
+					}
+				}
+				else if (_token.id != TokenId.Equal || _arguments.TryGetIdentifier(propertyOrFieldName, out _) || _arguments.TryGetParameters(propertyOrFieldName, out _))
+				{
+					SetTextPos(pos);
+					NextToken();
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:1272]: Aren't these two operations NOP?

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 13]
```diff
@@ -1207,41 +1207,134 @@ private Expression[] ParseArrayInitializerList()
...
+					ParseCollectionInitalizer(newType, pos, bindingList, actions, instance, allowCollectionInit);
+					return;
+				}
+				SetTextPos(pos);
 				NextToken();
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:1277]: Aren't these two operations NOP?

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs [hunk 14]
```diff
@@ -179,6 +180,148 @@ public void Array_multi_dimension_constructor()
...
+		[Test]
+		public void Ctor_NewMyClass_CanStillUseMemberSyntax()
+		{
+			var target = new Interpreter();
+			target.Reference(typeof(MyClassAdder));
+			target.Reference(typeof(MyClass));
+			Assert.AreEqual(
+				new MyClassAdder() { StrProp = string.Empty, MyArr = new[] { 1, 2, 3, 4, 5 }, IntField = int.MinValue },
+				target.Eval<MyClassAdder>("new MyClassAdder() {StrProp = string.Empty, MyArr = new int[] {1, 2, 3, 4, 5}, IntField = int.MinValue }"));
+		}
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ConstructorTest.cs:236]: Isn't this test case already handled by the previous test?

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs [hunk 15]
```diff
@@ -179,6 +180,148 @@ public void Array_multi_dimension_constructor()
...
+			target.Reference(typeof(MyClassAdder));
+			target.Reference(typeof(MyClass));
+			var strProp = new Parameter("StrProp", typeof(string).MakeByRefType(), "0");
+			var args = new object[]
+			{
+				strProp.Value
+			};
+			Assert.AreEqual(
+				new MyClassAdder() { { 1, 2, 3, 4, 5 }, "6", 7 },
+				target.Parse("new MyClassAdder(){{ 1, 2, 3, 4, 5},{StrProp = \"6\" },7}", strProp).Invoke(args));
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ConstructorTest.cs:218]: Nice test case, it took me a while to understand that this is not assigning the `StrProp` property, but instead assigning a value to the ref parameter `StrProp`!!

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs [hunk 16]
```diff
@@ -179,6 +180,148 @@ public void Array_multi_dimension_constructor()
...
+			Ctor_NewListGeneric<object, object>("string.Empty", "int.MinValue");
+		}
+
+		[Test]
+		public void Ctor_NewListCantFindAddMethod()
+		{
+			var target = new Interpreter();
+			target.Reference(typeof(System.Collections.Generic.List<>));
+			Assert.Throws<ParseException>(() => target.Eval<System.Collections.Generic.List<int>>("string.Empty"));
+			Assert.Throws<ParseException>(() => target.Eval<System.Collections.Generic.List<string>>("int.MaxValue"));
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ConstructorTest.cs:274]: I think you're missing the call to the List constructor here.

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs [hunk 17]
```diff
@@ -179,6 +180,148 @@ public void Array_multi_dimension_constructor()
...
+			//Create a random list of values to test.
+			var actual = new System.Collections.Generic.List<TObject>();
+			foreach (var v in items)
+			{
+				actual.Add(target.Eval<TObject>(v.Trim('}', '{')));
+			}
+			for (var min = 0; min < actual.Count; ++min)
+			{
+				for (var count = Math.Min(min, 1); count <= actual.Count - min; ++count)
+				{
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ConstructorTest.cs:292]: What's the purpose of the two loops? Wouldn't be enough to just concat all the provided items?

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs [hunk 18]
```diff
@@ -179,6 +180,148 @@ public void Array_multi_dimension_constructor()
...
+		[Test]
+		public void Ctor_NewListCantFindAddMethod()
+		{
+			var target = new Interpreter();
+			target.Reference(typeof(System.Collections.Generic.List<>));
+			Assert.Throws<ParseException>(() => target.Eval<System.Collections.Generic.List<int>>("string.Empty"));
+			Assert.Throws<ParseException>(() => target.Eval<System.Collections.Generic.List<string>>("int.MaxValue"));
+		}
+
+		public void Ctor_NewListGeneric<TList, TObject>(params string[] items)
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ConstructorTest.cs:277]: TList and TObject are always the same, so you could use a single generic argument.

File: test/DynamicExpresso.UnitTest/ParametersTest.cs [hunk 19]
```diff
@@ -332,6 +332,7 @@ public void When_lambda_is_invoked_input_parameters_must_follow_in_the_same_orde
 			Assert.AreEqual(4, lambda.Invoke(1, 5));
 		}
 
+
```
Comment (metoule) [inline test/DynamicExpresso.UnitTest/ParametersTest.cs:335]: The changes to this file can be reverted.

**Comment by metoule** [comment]
Sorry for the late review! The code looks good to me, the test cases as well, I still have a few questions, but nothing major.

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 12 above)
Comment (holdenmai) [reply src/DynamicExpresso.Core/Parsing/Parser.cs]: If I take that out, it ends up causing the (now fixed) Ctor_NewListCantFindAddMethod test to fail with "Expression Expected" instead of cannot find appropriate Add method

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 13 above)
Comment (holdenmai) [reply src/DynamicExpresso.Core/Parsing/Parser.cs:1277]: In the case of  being a Prop1 = "Value" in our parsing, the token is at =.  Calling this sets the token back to the Identifier Prop1.  
If this is not called, the NextToken call a couple lines below ends up advancing us to "Value"

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs (hunk 14 above)
Comment (holdenmai) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs:233]: Yes, missed in my splitting out of tests.

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs (hunk 15 above)
Comment (holdenmai) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs:218]: Yeah, this test was actually the one that caused me to raise #252 as I technically have no way of validating that it actually did the assign.

Comment (metoule) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs:218]: No, I don't think that's needed for now, it's really an edge case.

Comment (holdenmai) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs]: completed.

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs (hunk 16 above)
Comment (holdenmai) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs]: Correct

File: test/DynamicExpresso.UnitTest/ConstructorTest.cs (hunk 18 above)
Comment (holdenmai) [reply test/DynamicExpresso.UnitTest/ConstructorTest.cs]: Updated

**Review by davideicardi — APPROVED** [review]
